File: browse/port/harness/session_registry_oracle.go
//go:build ignore

// Command session_registry_oracle emits the stable Go session-registry contract.
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/danieljustus/symaira-browse/internal/daemon"
)

type sessionView struct {
	Name                   string `json:"name"`
	PID                    int    `json:"pid"`
	ActiveTabs             int    `json:"active_tabs"`
	UserDataDirBasename    string `json:"user_data_dir_basename"`
	UserDataPathIsClean    bool   `json:"user_data_path_is_clean"`
	BrowserContextID       string `json:"browser_context_id"`
	RefCount               int    `json:"ref_count"`
	Scope                  string `json:"scope"`
	OriginPath             string `json:"origin_path"`
	TimestampFormatValid   bool   `json:"timestamp_format_valid"`
	LastActivityNotEarlier bool   `json:"last_activity_not_earlier"`
}

type output struct {
	SchemaVersion      int            `json:"schema_version"`
	DefaultProfileRoot string         `json:"default_profile_root"`
	Sessions           []sessionView  `json:"sessions"`
	InvalidNameError   string         `json:"invalid_name_error"`
	MissingGetError    string         `json:"missing_get_error"`
	MissingTouchError  string         `json:"missing_touch_error"`
	EmptyRefError      string         `json:"empty_ref_error"`
	Reference          string         `json:"reference"`
	EnsureIdempotent   bool           `json:"ensure_idempotent"`
	Validation         []idValidation `json:"validation"`
	TouchAdvanced      bool           `json:"touch_advanced"`
	ClearedEntries     int            `json:"cleared_entries"`
	ProfilesPreserved  bool           `json:"profiles_preserved"`
}

type idValidation struct {
	Name     string `json:"name"`
	Accepted bool   `json:"accepted"`
}

func view(info daemon.SessionInfo) sessionView {
	started, startedErr := time.Parse(time.RFC3339Nano, info.StartedAt)
	activity, activityErr := time.Parse(time.RFC3339Nano, info.LastActivity)
	return sessionView{
		Name:                   info.Name,
		PID:                    info.PID,
		ActiveTabs:             info.ActiveTabs,
		UserDataDirBasename:    filepath.Base(info.UserDataDir),
		UserDataPathIsClean:    filepath.Clean(info.UserDataDir) == info.UserDataDir,
		BrowserContextID:       info.BrowserContextID,
		RefCount:               info.RefCount,
		Scope:                  info.Scope,
		OriginPath:             info.OriginPath,
		TimestampFormatValid:   startedErr == nil && activityErr == nil,
		LastActivityNotEarlier: startedErr == nil && activityErr == nil && !activity.Before(started),
	}
}

func main() {
	defaultRegistry := daemon.NewSessionRegistry(daemon.SessionRegistryOptions{})
	root, err := os.MkdirTemp("", "symbrowse-session-oracle-")
	if err != nil {
		panic(err)
	}
	defer os.RemoveAll(root)
	uncleanRoot := root + string(os.PathSeparator) + "unused" + string(os.PathSeparator) + ".."
	registry := daemon.NewSessionRegistry(daemon.SessionRegistryOptions{
		PID: 4242, UserDataRoot: uncleanRoot, Scope: "worktree", OriginPath: "/workspace/project",
	})
	_, invalidName := registry.Ensure("../escape")
	if invalidName == nil {
		panic("invalid session name was accepted")
	}
	if _, err := registry.Ensure("beta"); err != nil {
		panic(err)
	}
	alpha, err := registry.Ensure("alpha")
	if err != nil {
		panic(err)
	}
	again, err := registry.Ensure("alpha")
	if err != nil {
		panic(err)
	}
	if err := registry.SetActiveTabs("alpha", 3); err != nil {
		panic(err)
	}
	if err := registry.SetRef("alpha", "selector", "@element-1"); err != nil {
		panic(err)
	}
	ref, err := registry.Ref("alpha", "selector")
	if err != nil {
		panic(err)
	}
	beforeTouch, err := registry.Get("alpha")
	if err != nil {
		panic(err)
	}
	time.Sleep(2 * time.Millisecond)
	if err := registry.Touch("alpha"); err != nil {
		panic(err)
	}
	afterTouch, err := registry.Get("alpha")
	if err != nil {
		panic(err)
	}

	validationRegistry := daemon.NewSessionRegistry(daemon.SessionRegistryOptions{UserDataRoot: filepath.Join(root, "validation")})
	validation := make([]idValidation, 0, 7)
	for _, name := range []string{"", strings.Repeat("x", 65), "../escape", "ümlaut", "has space"} {
		_, err := validationRegistry.Ensure(name)
		validation = append(validation, idValidation{Name: name, Accepted: err == nil})
	}
	for _, name := range []string{"a", strings.Repeat("x", 64)} {
		_, err := validationRegistry.Ensure(name)
		validation = append(validation, idValidation{Name: name, Accepted: err == nil})
	}
	data := registry.ListData()
	views := make([]sessionView, 0, len(data.Sessions))
	for _, info := range data.Sessions {
		views = append(views, view(info))
	}
	_, missingGet := registry.Get("missing")
	missingTouch := registry.Touch("missing")
	emptyRef := registry.SetRef("alpha", "", "value")
	registry.Clear()
	profileExists := true
	for _, name := range []string{"alpha", "beta"} {
		if _, err := os.Stat(filepath.Join(root, name)); err != nil {
			profileExists = false
		}
	}
	if !profileExists {
		panic("session registry clear removed profile directories")
	}
	result := output{
		SchemaVersion: data.SchemaVersion, DefaultProfileRoot: defaultRegistry.UserDataRoot(), Sessions: views,
		InvalidNameError: invalidName.Error(),
		MissingGetError:  missingGet.Error(), MissingTouchError: missingTouch.Error(),
		EmptyRefError: emptyRef.Error(), Reference: ref, EnsureIdempotent: again == alpha,
		Validation: validation, TouchAdvanced: afterTouch.LastActivity != beforeTouch.LastActivity,
		ClearedEntries: len(registry.List()), ProfilesPreserved: profileExists,
	}
	if err := json.NewEncoder(os.Stdout).Encode(result); err != nil {
		panic(fmt.Errorf("encode oracle: %w", err))
	}
}

File: browse/crates/symbrowse-daemon/examples/session_registry_oracle.rs
use std::{
    fs,
    path::PathBuf,
    time::{SystemTime, UNIX_EPOCH},
};

use symbrowse_daemon::{SessionRegistry, SessionRegistryOptions};

fn timestamp_state(started_at: &str, last_activity: &str) -> (bool, bool) {
    let started =
        time::OffsetDateTime::parse(started_at, &time::format_description::well_known::Rfc3339);
    let activity = time::OffsetDateTime::parse(
        last_activity,
        &time::format_description::well_known::Rfc3339,
    );
    match (started, activity) {
        (Ok(started), Ok(activity)) => (true, activity >= started),
        _ => (false, false),
    }
}

fn main() {
    let nonce = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("system time after epoch")
        .as_nanos();
    let root = std::env::temp_dir().join(format!(
        "symbrowse-session-oracle-{}-{nonce}",
        std::process::id()
    ));
    fs::create_dir(&root).expect("create isolated registry root");
    let unclean_root = root.join("unused").join("..");
    let registry = SessionRegistry::new(SessionRegistryOptions {
        user_data_root: unclean_root,
        pid: 4242,
        scope: "worktree".into(),
        origin_path: "/workspace/project".into(),
    });

    let invalid_name_error = registry
        .ensure("../escape")
        .expect_err("invalid session name rejected")
        .to_string();
    registry.ensure("beta").expect("ensure beta");
    let alpha = registry.ensure("alpha").expect("ensure alpha");
    let again = registry.ensure("alpha").expect("ensure alpha again");
    let ensure_idempotent = again.name == alpha.name
        && again.started_at == alpha.started_at
        && again.user_data_dir == alpha.user_data_dir
        && again.browser_context_id == alpha.browser_context_id;
    registry
        .set_active_tabs("alpha", 3)
        .expect("set active tabs");
    registry
        .set_ref("alpha", "selector", "@element-1")
        .expect("set ref");
    let reference = registry.reference("alpha", "selector").expect("read ref");
    registry.touch("alpha").expect("touch alpha");

    let data = registry.list_data();
    let sessions: Vec<_> = data
        .sessions
        .iter()
        .map(|info| {
            let (timestamp_format_valid, last_activity_not_earlier) =
                timestamp_state(&info.started_at, &info.last_activity);
            serde_json::json!({
                "name": info.name,
                "pid": info.pid,
                "active_tabs": info.active_tabs,
                "user_data_dir_basename": PathBuf::from(&info.user_data_dir)
                    .file_name().and_then(|name| name.to_str()).unwrap_or_default(),
                "user_data_path_is_clean": PathBuf::from(&info.user_data_dir)
                    == registry.user_data_root().join(&info.name),
                "browser_context_id": info.browser_context_id,
                "ref_count": info.ref_count,
                "scope": info.scope,
                "origin_path": info.origin_path,
                "timestamp_format_valid": timestamp_format_valid,
                "last_activity_not_earlier": last_activity_not_earlier,
            })
        })
        .collect();
    let missing_get_error = registry.get("missing").unwrap_err().to_string();
    let missing_touch_error = registry.touch("missing").unwrap_err().to_string();
    let empty_ref_error = registry
        .set_ref("alpha", "", "value")
        .unwrap_err()
        .to_string();
    registry.clear();
    let profiles_preserved = root.join("alpha").is_dir() && root.join("beta").is_dir();
    assert!(
        profiles_preserved,
        "registry clear must preserve browser profile directories"
    );
    let result = serde_json::json!({
        "schema_version": data.schema_version,
        "sessions": sessions,
        "invalid_name_error": invalid_name_error,
        "missing_get_error": missing_get_error,
        "missing_touch_error": missing_touch_error,
        "empty_ref_error": empty_ref_error,
        "reference": reference,
        "ensure_idempotent": ensure_idempotent,
        "cleared_entries": registry.list().len(),
        "profiles_preserved": profiles_preserved,
    });
    println!("{result}");
    fs::remove_dir_all(root).expect("remove only this oracle's temporary registry root");
}
